### PR TITLE
Problem (Fix #1121): Too tedious to create separated c apis for all client features

### LINKED ISSUE #1121: Problem: Too tedious to create separated c apis for all client features
Separated from #1115.

We can classify c apis into two categories:
- Stateless, these are pure functions, and we should provide them with normal c APIs, they can be embedded into user's code perfectly.
  - keypair derivation
  - transaction signing
  - fee policy
  - UTxO selection algorithm implemented in a stateless way
- Stateful features, there are features involves storage state and blocking network IO.
  These APIs should be taken extra care when used in UI thread or event-driven environment.
  - Wallet syncing.
  - Transaction building and broadcasting.
  - Basically most of the features in `client-rpc`.

I propose to expose the stateful features with a common JSON API:

```rust
fn jsonrpc_call(storage_path, tendermint_rpc_url, network_id, json_request) -> json_response
```

> It might be expanded a little bit as we investigate deeper, but the number of APIs should be a handful.

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- create `jsonrpc_call` api to encapulate all commands of `client-rpc`.

### CODE REVIEW

**Comment by yihuang** [comment]
Python test:
```python
import json
from ctypes import *

lib = cdll.LoadLibrary('../target/debug/libcro_clib.dylib')
lib.jsonrpc_call.argtypes = (c_char_p, c_char_p, c_char, c_char_p, c_char_p, c_size_t)

storage_path = './data/node0/wallet'.encode()
wsurl = 'ws://127.0.0.1:26657/websocket'.encode()
req = json.dumps({'jsonrpc': '2.0', 'method': 'wallet_list', 'params': [], 'id': 1}).encode()
output = create_string_buffer(1024)
retcode = lib.jsonrpc_call(storage_path, wsurl, b'\xAB', req, output, len(output))
assert retcode == 0, output.value.decode()
print(json.loads(output.value))
# {'jsonrpc': '2.0', 'result': ['Default'], 'id': 1}
```

**Comment by tomtau** [comment]
bors try

**Comment by yihuang** [comment]
One known inefficiency with this API is it open/close the DB, also spawn/stop the auto-synchronizer thread for each API call. We can further optimize that if we agree on the overall direction.

File: client-core/src/synchronizer.rs [hunk 1]
```diff
@@ -19,6 +20,9 @@ type WalletPassphrase = SecKey;
 pub struct PollingSynchronizer {
     wallets: Arc<Mutex<BTreeMap<WalletName, WalletPassphrase>>>,
```
Comment (leejw51) [inline client-core/src/synchronizer.rs]: how about just using one mutex , for PollingSynchronizer

File: client-core/src/synchronizer.rs [hunk 2]
```diff
@@ -19,6 +20,9 @@ type WalletPassphrase = SecKey;
...
     progress: Arc<Mutex<SynchronizerProgress>>,
+    should_run: Arc<AtomicBool>,
+    sync_thread: Option<thread::JoinHandle<()>>,
```
Comment (leejw51crypto) [inline client-core/src/synchronizer.rs]: how about using one mutx for wallets, progress, should_run

File: client-core/src/synchronizer.rs [hunk 3]
```diff
@@ -19,6 +20,9 @@ type WalletPassphrase = SecKey;
 pub struct PollingSynchronizer {
     wallets: Arc<Mutex<BTreeMap<WalletName, WalletPassphrase>>>,
     progress: Arc<Mutex<SynchronizerProgress>>,
+    should_run: Arc<AtomicBool>,
```
Comment (leejw51crypto) [inline client-core/src/synchronizer.rs]: houw about just using bool?

File: client-core/src/synchronizer.rs [hunk 4]
```diff
@@ -45,30 +49,37 @@ impl PollingSynchronizer {
...
+                    .clone();
+
+                for (name, enckey) in wallets_to_synchronize.iter() {
+                    let result = WalletSyncer::with_obfuscation_config(
+                        config.clone(),
+                        Some(sender.clone()),
+                        name.clone(),
+                        enckey.clone(),
+                    )
+                    .and_then(|syncer| syncer.sync());
```
Comment (leejw51crypto) [inline client-core/src/synchronizer.rs]: in testnet, this sync will take many hours.
maybe days for slow network

so i think for a few wallets,
each thread for wallet ?

or one block data, multiplexing,
but that's beyond this PR

File: client-core/src/synchronizer.rs [hunk 5]
```diff
@@ -45,30 +49,37 @@ impl PollingSynchronizer {
...
-                    config.clone(),
-                    Some(sender.clone()),
-                    name.clone(),
-                    enckey.clone(),
-                )
-                .and_then(|syncer| syncer.sync());
-                if let Err(e) = result {
-                    log::error!("Error while synchronizing wallet [{}]: {:?}", name, e);
+        self.sync_thread = Some(thread::spawn(move || {
+            while should_run.load(Ordering::Relaxed) {
```
Comment (leejw51crypto) [inline client-core/src/synchronizer.rs]: while should_run {

File: cro-clib/src/jsonrpc.rs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+}
+
+fn do_jsonrpc_call(
+    storage_dir: &str,
+    websocket_url: &str,
+    network_id: u8,
+    json_request: &str,
+) -> Result<String> {
+    let mut io = IoHandler::new();
+    let storage = SledStorage::new(storage_dir)?;
```
Comment (leejw51crypto) [inline cro-clib/src/jsonrpc.rs]: how de-coupling storage?
so that user supply other database for storage

File: cro-clib/src/jsonrpc.rs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        .err_kind(ErrorKind::InvalidInput, || {
+            "stateful command execute failed"
+        })
+}
+
+/// # Safety
+///
+/// Should not be called with null pointers.
+#[no_mangle]
+pub unsafe extern "C" fn jsonrpc_call(
```
Comment (leejw51crypto) [inline cro-clib/src/jsonrpc.rs]: this makes temporary rpc server,
so in this context, io port , should be found by using socket api (bind)

but rpc server everytime is very slow.
i suggest separating into two pars
1. initialize rpc server  <- initialize_json_rpc
2. use rpc server via rpc <- json_rpc_call

File: client-core/src/synchronizer.rs (hunk 3 above)
Comment (yihuang) [reply client-core/src/synchronizer.rs]: It’s shared between threads, so atomic is needed.

File: client-core/src/synchronizer.rs [hunk 8]
```diff
@@ -106,7 +117,18 @@ impl PollingSynchronizer {
...
             }
-        });
+        }));
+    }
+
+    /// Stop the auto synchronizer thread
+    pub fn stop(&mut self) {
+        self.should_run.store(false, Ordering::Relaxed);
+        if let Some(thread) = self.sync_thread.take() {
+            thread.join().unwrap();
```
Comment (leejw51crypto) [inline client-core/src/synchronizer.rs]: expect

File: client-core/src/synchronizer.rs (hunk 2 above)
Comment (yihuang) [reply client-core/src/synchronizer.rs]: No, they are for different purposes.

File: client-core/src/synchronizer.rs (hunk 4 above)
Comment (yihuang) [reply client-core/src/synchronizer.rs]: It's already recorded in #1060

File: client-core/src/synchronizer.rs (hunk 5 above)
Comment (yihuang) [reply client-core/src/synchronizer.rs]: https://github.com/crypto-com/chain/pull/1122#discussion_r383067007

File: cro-clib/src/jsonrpc.rs (hunk 6 above)
Comment (yihuang) [reply cro-clib/src/jsonrpc.rs]: That's a potential feature for the future.

File: cro-clib/src/jsonrpc.rs (hunk 7 above)
Comment (yihuang) [reply cro-clib/src/jsonrpc.rs]: No server gets started, directly calls the RPC functions, not go through the socket.

File: cro-clib/chain.h [hunk 9]
```diff
@@ -437,3 +437,15 @@ CroResult cro_withdraw(const char *tenermint_url_string,
                        int32_t viewkey_count,
                        uint8_t *output,
                        uint32_t *output_length);
+
+/**
+ * # Safety
+ *
+ * Should not be called with null pointers
+ */
```
Comment (leejw51crypto) [inline cro-clib/chain.h]: add some example for storage, tendermint url

File: cro-clib/chain.h [hunk 10]
```diff
@@ -437,3 +437,15 @@ CroResult cro_withdraw(const char *tenermint_url_string,
...
+CroResult jsonrpc_call(const char *storage_dir,
```
Comment (leejw51crypto) [inline cro-clib/chain.h]: how about json_rpc

Comment (yihuang) [reply cro-clib/chain.h]: I'll add a `cro_` prefix.

**Comment by yihuang** [comment]
> @yihuang Can you add more context in this PRs description? Why is it too tedious to create C APIs (root cause) and how does this PR tries to solve that root cause?

It's an alternative approach to current c API, with clear advantages and disadvantages if works for the user:
- Much smaller code to maintain, current cro-clib already has 1591 lines of rust code with incomplete features, this approach only need 181 lines of rust code to provide complete client features, and the consistency with `client-rpc` is kept automatically.
  ```
  $ cat cro-clib/src/*.rs|wc -l
   1591
  $ git checkout issue1121
  $ cat cro-clib/src/jsonrpc.rs|wc -l
    181
  ```

Disadvantages are relative inefficiency because of JSON serialization/deserialization, which I mentions in the previous comments.

File: cro-clib/chain.h (hunk 9 above)
Comment (yihuang) [reply cro-clib/chain.h]: Done, added a c example in the doc.

File: cro-clib/chain.h (hunk 10 above)
Comment (yihuang) [reply cro-clib/chain.h]: Done

File: client-core/src/synchronizer.rs (hunk 8 above)
Comment (yihuang) [reply client-core/src/synchronizer.rs]: Done

**Comment by leejw51crypto** [comment]
i think both approaches are fine,
for c api, all control and ownership belongs to the user.
especially for wallet

**Comment by yihuang** [comment]
> may be ok -- one important thing is that for "sync", it'll be pretty bad for the caller, as that may be long-running and blocking (and you don't have any feedback when that is supposed to finish ... or whether it got stuck or something).
> I suppose it could also take a callback function pointer that could be repeatedly called with progress?

Yes, there are quite a few things to be improved, the current one is like a minimal working prototype:
- Separate the open/execute/close, so we can share opened resources between multiple calls.
- There are logging codes in `client-rpc`, for an API library, the user probably not expect logs inside.
- For the sync progress report, we can have some common stream output support, something like this?:
  ```
  let stream_id = jsonrpc_call('add wallet to sync');
  jsonrpc_stream_watch(stream_id, callback);
  ```

**Comment by leejw51crypto** [comment]
callback or polling is necessary for really long request

for sync, i'm thinking similar thing with yi huang
1. give basic apis
-open, close, get block
- user owns flow control, manages sync by himeself

2. automatic state machine which has its own thread
-open, close, callback or polling
-storage all take care of the state machine

3. simple method <- current json rpc
provide some polling or callback to get progress
just like light-weight client-rpc

**Comment by tomtau** [comment]
> Yes, there are quite a few things to be improved,

@yihuang @leejw51crypto so shall this PR be merged? There can be separate issues on the improvements?

**Comment by leejw51crypto** [comment]
i think, we can merge this pr,
and improve by another pr

**Comment by tomtau** [comment]
@yihuang can you squash the commits?

**Comment by yihuang** [comment]
> @yihuang can you squash the commits?

Done

**Pull request merged**